### DISCUSSION TITLE
Querying multiple MPD instances at once and displaying results in a common user interface view

### DISCUSSION THREAD

**Discussion by m7a** [MusicPlayerDaemon/MPD#2029] [Q&A] [answered]
Hello MPD community,

I am working on creating an MPD client that can connect to two MPD servers (on different machines) at once. The different servers are expected to contain different, but similar music directories. E.g. a remote machine could have an older copy of the same music database as the current machine execept that all songs have been resampled to a different sample rate.

The client is intended to allow working with this setup by visually indicating (e.g. with colors) when a song is not available on all clients, but in general it is expected to show all songs that are available on at least one client.

I am currently thinking about how to implement a list of all songs across all of the servers sorted by a common criterion e.g. lexicographically by artist. For an individual MPD instance, this is easy to do using `find` with a `WINDOW` related to the UI viewport.

For multiple servers, it is more complicated, though: I cannot just assume that the same `WINDOW` would be applicable to multiple servers, because the lists of songs returned might not correspond to the same lexicographic range across the servers.

I currently consider two solutions

 - (a) Download the entire (or a very large portion of) the music database from all servers and assemble the result list in the client. The documentation warns against doing this and I think in this case there might be a way around it, but I am not sure whether it is worth implementing (see next point). The advantage would be that it is easier to implement it correctly and also the UI latency for scrolling would be minimal (since no interaction with the MPD servers is needed for scrolling etc.)
 - (b) Incrementally query: I'd start working with a WINDOW that is comparatively small (say similar to the viewport size ... 20 entries or something). I'd query all servers and as long as the query results are _all before_ or _all after_ the ones from an other server, I'd adjust the `WINDOW` (to query 20 entries later / 20 entries earlier -- or maybe some form of binary search...) until having identified the appropriate WINDOW ranges accross the servers to populate a common song list. While it seems to be plausible to do this from a theoretical point of view, I can imagine that this could have quite the latency from the UI side and may also be prone to endless loops in corner cases.

Is there any alternative to these options for my use case?

Which solution would seem more sensible from an MPD point of view?

Thanks in advance,
Linux-Fan (@m7a)

**Comment by MaxKellermann** [comment, marked as answer]
General rule: at any given time, you should only query the part of the MPD database that is needed. If you have a million songs, it never makes sense to query all the tags from a million songs - you can't display them anyway.
If you do display many songs, query only the tags that you will use for that display; use the `tagtypes` command to tell MPD to omit tag values from the response that your client will never show.
If you don't actually want to display a list of songs, but a list of artists, don't use the `find` command but instead `list artist`. Let the server aggregate the result for you.

> I am currently thinking about how to implement a list of all songs across all of the servers sorted by a common criterion e.g. lexicographically by artist. For an individual MPD instance, this is easy to do using `find` with a `WINDOW` related to the UI viewport.

Do `list artist` on all instances, merge the results. There's no `window` for `list`, because the response is usually small enough.

**Reply by m7a** [reply]
Thank you very much for the quick response. I shall try to avoid downloading the entire database and see how well it works :smile: The hint with querying the list of artists directly is definitely a good one.
